gpbft/gpbft.go

```diff
@@ -482,31 +489,35 @@ func (i *instance) beginQuality() error {
 
 // Attempts to end the QUALITY phase and begin PREPARE based on current state.
 func (i *instance) tryQuality() error {
-	if i.phase != QUALITY_PHASE {
-		return fmt.Errorf("unexpected phase %s, expected %s", i.phase, QUALITY_PHASE)
-	}
-	// Wait either for a strong quorum that agree on our proposal,
-	// or for the timeout to expire.
-	foundQuorum := i.quality.HasStrongQuorumFor(i.proposal.Key())
-	timeoutExpired := atOrAfter(i.participant.host.Time(), i.phaseTimeout)
-
-	if foundQuorum {
-		// Keep current proposal.
-	} else if timeoutExpired {
-		strongQuora := i.quality.ListStrongQuorumValues()
-		i.proposal = findFirstPrefixOf(i.proposal, strongQuora)
-	}
+	// Note that QUALITY messages are accepted at any round/phase.
+	if i.phase == QUALITY_PHASE {
+		// Wait either for a strong quorum that agree on input, or for the timeout to
+		// expire.
+		foundQuorum := i.quality.HasStrongQuorumFor(i.input.Key())
+		timeoutExpired := atOrAfter(i.participant.host.Time(), i.phaseTimeout)
+
+		if foundQuorum {
+			// Keep current proposal.
+		} else if timeoutExpired {
+			// Change proposal to the longest prefix of input with strong quorum.
+			i.proposal = i.quality.FindStrongQuorumValueForLongestPrefixOf(i.input)
+		}
```

Comment:
Wouldn't this be the same, regardless of whether our proposal won or just its prefix?

Reply:
hmm, but we probably want to keep the behaviour of: move on when our proposal wins or timeout explores, instead of changing it to move on when any proposal wins

Reply:
Sorry, I don't follow. Same in what sense?

Reply:
The behaviour of the case where current phase is QUALITY is unchanged relative to `main` unless I have missed something.